### PR TITLE
[OSS/ShardedDDP] making APIs more private

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Only keeps public what's reasonably necessary

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
Benchnmarks are unrelated, fixed by https://github.com/facebookresearch/fairscale/pull/581

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -517,20 +454,75 @@ def add_param_group(self, param_group: dict) -> None:
             # Update the bucketing strategy accordingly
             self._setup_flat_buffers()
 
+    @property
+    def _local_params(self) -> List[torch.Tensor]:
+        """ Iterable which goes through the parameters that this rank owns
+        """
```
Comment (anj-s) [inline fairscale/optim/oss.py:460]: should this be on the previous line given this is a single line comment? 
nit: period at the end of the comment.

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -517,20 +454,75 @@ def add_param_group(self, param_group: dict) -> None:
...
+                    ]
+                )
+            )
+
+        # Make sure that the iterator is not consumed, only expose a copy
+        return self.__local_params
+
+    @property
+    def _param_to_index(self) -> Dict[int, int]:
+        """ Hash table in between parameter indices in the global optimizer scheme, and the actual params
```
Comment (anj-s) [inline fairscale/optim/oss.py:476]: same comment as above.

File: fairscale/optim/oss.py [hunk 3]
```diff
@@ -517,20 +454,75 @@ def add_param_group(self, param_group: dict) -> None:
...
+    @property
+    def _param_to_rank(self) -> Dict[torch.Tensor, int]:
+        """param to data parallel rank"""
+        if len(self.__param_rank) == 0:
+            for rank, param_groups in enumerate(self.partition_parameters()):
+                for param_group in param_groups:
+                    for param in param_group["params"]:
+                        self.__param_rank[param] = rank
+
+            logging.debug("ZeRO: Parameters dispatched to ranks %s " % list(self.__param_rank.values()))
```
Comment (anj-s) [inline fairscale/optim/oss.py:514]: rename to "OSS" or "SDP" instead of ZeRO?

File: fairscale/optim/oss.py [hunk 4]
```diff
@@ -517,20 +454,75 @@ def add_param_group(self, param_group: dict) -> None:
...
+            # Sort param_lists by size
+            for device in self.__per_device_params.keys():
+                for rank_params in self.__per_device_params[device]:
+                    rank_params.sort(key=lambda x: x.numel())
+
+        return self.__per_device_params
+
+    @property
+    def _param_to_rank(self) -> Dict[torch.Tensor, int]:
+        """param to data parallel rank"""
```
Comment (anj-s) [inline fairscale/optim/oss.py:507]: perhaps a slightly more verbose string that mentions which "param" would be good.

File: fairscale/optim/utils.py [hunk 5]
```diff
@@ -18,6 +18,14 @@ def __init__(self, handle: Any, callback: Optional[Callable]) -> None:
         self.callback = callback
 
 
+def get_global_rank(group: Any, rank: int) -> int:
+    if group is dist.group.WORLD:
+        return rank
+    else:
```
Comment (anj-s) [inline fairscale/optim/utils.py:24]: redundant "else"clause I think. You can just say
```
if if group is dist.group.WORLD:
  return rank

return dist.distributed_c10d._get_global_rank(group, rank)
```
WDYT?

File: fairscale/optim/oss.py [hunk 6]
```diff
@@ -89,11 +89,11 @@ def __init__(
         self.in_super_constructor = False
 
         # Partition information. lazy evaluation, computed when requested
-        self._per_device_params: Dict[torch.device, List[List[Parameter]]] = OrderedDict()  # device, rank, params
-        self._param_rank: Dict[torch.Tensor, int] = {}
+        self.__per_device_params: Dict[torch.device, List[List[Parameter]]] = OrderedDict()  # device, rank, params
```
Comment (anj-s) [inline fairscale/optim/oss.py:92]: Nice! I like this :)

File: fairscale/optim/oss.py (hunk 4 above)
Comment (blefaudeux) [reply fairscale/optim/oss.py]: I tried to explain a little more, this applies to all params actually :)

**Review by anj-s — APPROVED** [review]

**Pull request merged**